gpbft/validator.go

```diff
@@ -43,154 +73,261 @@ func newValidator(host Host, cp CommitteeProvider, progress Progress, cache *cac
+	}
+
+	// Check the message again relatively to the current progress in case it's become
+	// irrelevant. This check is inexpensive, so we do it again even though it was already
+	// done in partial validation.
+	if err := v.validateByProgress(pmsg.GMessage); err != nil {
+		return nil, err
+	}
+
+	// If the key is zero, then the vote value must be zero, along with justification
+	// vote value if justification is present.
+	justified := pmsg.Justification != nil
+	if pmsg.VoteValueKey.IsZero() {
+		if !pmsg.Vote.Value.IsZero() {
+			return nil, fmt.Errorf("unexpected non-zero value for zero vote value key: %w", ErrValidationInvalid)
+		}
+		if justified && !pmsg.Justification.Vote.Value.IsZero() {
+			return nil, fmt.Errorf("unexpected non-zero justification value for zero vote value key: %w", ErrValidationInvalid)
 		}
-	default:
-		// Message belongs to an instance older than the previous instance.
-		return nil, ErrValidationTooOld
 	}
+	if justified {
+		// Abbreviated version of the expectation map from the full validator.
+		expectations := map[Phase]map[Phase]*ECChain{
+			CONVERGE_PHASE: {
+				COMMIT_PHASE:  &ECChain{},
+				PREPARE_PHASE: pmsg.Vote.Value,
+			},
+			PREPARE_PHASE: {
+				COMMIT_PHASE:  &ECChain{},
+				PREPARE_PHASE: pmsg.Vote.Value,
+			},
+			COMMIT_PHASE: {
+				PREPARE_PHASE: pmsg.Vote.Value,
+			},
+			DECIDE_PHASE: {
+				COMMIT_PHASE: pmsg.Vote.Value,
+			},
+		}
+		if expectedPhases, ok := expectations[pmsg.Vote.Phase]; ok {
+			if expectedValue, ok := expectedPhases[pmsg.Justification.Vote.Phase]; ok {
+				if !pmsg.Justification.Vote.Value.Eq(expectedValue) {
+					return nil, fmt.Errorf("message %v has justification for a different value: %v: %w", pvmsg, pmsg.Justification.Vote.Value, ErrValidationInvalid)
+				}
+			} else {
+				return nil, fmt.Errorf("message %v has justification with unexpected phase: %v: %w", pvmsg, pmsg.Justification.Vote.Phase, ErrValidationInvalid)
+			}
+		} else {
+			return nil, fmt.Errorf("message %v has unexpected phase for justification: %w", pvmsg, ErrValidationInvalid)
+		}
+	}
+	return &validatedMessage{msg: pmsg.GMessage}, nil
+}
 
-	var buf bytes.Buffer
-	var cacheMessage bool
-	if err := msg.MarshalCBOR(&buf); err != nil {
-		log.Errorw("failed to marshal message for caching", "err", err)
-	} else if alreadyValidated, err := v.cache.Contains(msg.Vote.Instance, messageCacheNamespace, buf.Bytes()); err != nil {
-		log.Errorw("failed to check already validated messages", "err", err)
+func (v *cachingValidator) ValidateMessageWithVoteValueKey(ctx context.Context, valueKey *ECChainKey, msg *GMessage) (_err error) {
+	if msg == nil {
+		return ErrValidationInvalid
+	}
+
+	if err := v.validateByProgress(msg); err != nil {
+		return err
+	}
+
+	partial := valueKey != nil
+	cacheNamespace := validationNamespaces.message(partial)
+	alreadyValidated, cacheKey, err := v.isAlreadyValidated(msg.Vote.Instance, cacheNamespace, msg)
```

Comment:
This seems a bit iffy for me. IDK if I'm right.
The validation cache only keys on the `msg`, but ignores the `valueKey`; this would result in messages passing the validation with the wrong key.

Reply:
Solution is to either cache on `valueKey` as well, or to compare the `valueKey` before the cache check.
Computing `ECChain.Key()` should be free in many cases as it will be internally cached.